### PR TITLE
ROM Integrity test

### PR DESCRIPTION
**Pull request by korran**
As usual, I recommend reviewing each commit independently.

### CODE REVIEW

**Review by JohnTraverAmd — DISMISSED** [review]

File: drivers/src/sha256.rs [hunk 1]
```diff
@@ -215,9 +244,6 @@ impl Sha256 {
             sha256.ctrl().write(|w| w.mode(true).init(false).next(true));
         }
```
Comment (mhatrevi) [inline drivers/src/sha256.rs:245]: Why is this removed?

**Review by mhatrevi — CHANGES_REQUESTED** [review]

**Comment by rusty1968** [comment]
@korran , is this still going into ROM? @JohnTraverAmd  made this a requirement for runtime so perhaps the entry point for invoking the test could have an environment parameter that is decoupled from ROM.

This is a pre-requisite for the self-test pr for runtime.

**Comment by korran** [comment]
> @korran , is this still going into ROM? @JohnTraverAmd made this a requirement for runtime so perhaps the entry point for invoking the test could have an environment parameter that is decoupled from ROM.
> 
> This is a pre-requisite for the self-test pr for runtime.

Yes, this needs to go into the ROM. I assume the entry point for "run all the FIPS tests" will include running this test, which the runtime firmware can call on-demand.

There's no need to do an integrity test of the FMC or runtime firmware at boot because that's done as part of the ROM validating the image. I suppose there is still a need for re-validating the FMC or runtime firmware on request, but I don't think that is related to this ROM self-test; that logic should be running the full FMC / runtime verification routine that includes parsing and validating the image headers.

File: drivers/src/sha256.rs (hunk 1 above)
Comment (mhatrevi) [reply drivers/src/sha256.rs:245]: Since this change is unrelated to the ROM integrity testing, can we make this as a separate change?

Comment (mhatrevi) [reply drivers/src/sha256.rs:245]: For a multi-block case, how do we ensure that the crypto engine is not processing the data in the block registers? digest_block simply copies the data to the block registers and then calls digest_op, which then checks for status.ready. Previously, digest_op would have only returned when the processing was complete.

**Comment by rusty1968** [comment]
> > @korran , is this still going into ROM? @JohnTraverAmd made this a requirement for runtime so perhaps the entry point for invoking the test could have an environment parameter that is decoupled from ROM.
> > This is a pre-requisite for the self-test pr for runtime.
> 
> Yes, this needs to go into the ROM. I assume the entry point for "run all the ROM FIPS tests" will include running this test, which the runtime firmware can call on-demand.
> 
> There's no need to do an integrity test of the FMC or runtime firmware at boot because that's done as part of the ROM validating the image. I suppose there is still a need for re-validating the FMC or runtime firmware on request, but I don't think that is related to this ROM self-test; that logic should be running the full FMC / runtime verification routine that includes parsing and validating the image headers. Unfortunately because that requires doing the SHA checksum against ICCM, it won't be the exact same logic as the ROM uses (which uses the mailbox sha384acc hardware).

@JohnTraverAmd had specified that RT would invoke ROM integrity tests as part of command handling : i.e. linking the ROM integrity test code and invoking it in context of RT.

File: drivers/src/sha256.rs (hunk 1 above)
Comment (korran) [reply drivers/src/sha256.rs:245]: I asked Michael:

> Kor: I'm trying to stream data to the sha256 peripheral as fast as possible, do I need to wait for the ready bit before writing to the SHA256_BLOCK register, or do I only have to check that before writing the command? (edited) 
> 
> Michael: I believe init/next commands will latch the data in BLOCK. You can begin writing the next block while the command is running setting next again only after ready
> 
> Kor: ok, that's my interpretation of the RTL as well
> 
> Michael: That's how I do it for sha accelerator and any tests we've written. I only check ready prior to starting the next command

You can see it grabbing the entire register immediately after the command is received:

https://github.com/chipsalliance/caliptra-rtl/blob/main/src/sha256/rtl/sha256_w_mem.v#L209

Comment (vsonims) [reply drivers/src/sha256.rs:245]: Is this really needed? Can we continue to use the old interface. How much does this improve the time by?

**Review by vsonims — DISMISSED** [review]

File: drivers/src/sha256.rs (hunk 1 above)
Comment (korran) [reply drivers/src/sha256.rs:245]: digest_blocks_raw() is definitely required, as it is undefined behavior to create a slice reference to address zero. No way around that one.

Removing the wait for status::valid saves 45,483 cycles, due to the additional pipelining.

Comment (vsonims) [reply drivers/src/sha256.rs:245]: Is ~45000 a big win that warrents such large changes this late?

Comment (korran) [reply drivers/src/sha256.rs:245]: I honestly don't know, since we don't have much guidance on performance. This does delay how long it takes the ROM to be ready for the first measurements.

I did the optimization because I wanted to collect perf numbers with the optimal implementation when I wrote this 2 months ago.

If you're not comfortable, I can revert.

Comment (korran) [reply drivers/src/sha256.rs]: Ok, in the interest of getting this PR in sooner, I've reverted the validity check optimization.

**Review by mhatrevi — APPROVED** [review]
